### ISSUE TITLE
How to format domain regex?

### ISSUE THREAD

**Issue by mikewindow** [objective-see/LuLu#241]
Do the caret and dollar sign need to be included when entering a regex value into the the "Remote address/domain" field? Should the value be `^api\.netscape\.com` or `netscape\.com`?

<img width="504" alt="screenshot" src="https://user-images.githubusercontent.com/61528978/97646582-46d20c80-1a0d-11eb-8ac7-5a9f5696813c.png">

**Comment by tomtastic** [comment]
I'm also wondering how these work, I tried creating a regex to match all subdomains with ".*\.protonmail\.ch", but as the screenshot shows it fails to match it...

I'm not sure why the LuLu prompt didn't do a reverse lookup on the IP address, is this why the matching logic seems to have failed?

Also would be nice if LuLu showed whois netname information if it can't do a PTR lookup for whatever reason..
.
![image](https://user-images.githubusercontent.com/302127/99146061-73706000-266c-11eb-94db-2b7c5f903f0f.png)

**Comment by ohuf** [comment]
Hi there.
Is the "regex" feature working at all?
I wanted to allow access to my local IP range (192.168.67.0/24), so I added a destination regex of 
192\.168\.1\..*
(every host in the subnet) but it's not working!

Would be nice to have a few examples for the regex feature!

Cheers!

**Comment by tomtastic** [comment]
I look forward to this getting fixed, as I had to stop using LuLu due to to this issue. Not being able to resolve IPs to the domains being contacted made for a horrible user experience, the regex issue compounds this.

**Comment by xanoni** [comment]
Overall, I think that Lulu is not doing a great job at explaining what the expected input is when defining rules. There should be examples or tooltips in the app, or at least a comprehensive documentation. Currently it seems you can't know for sure without looking at the code. 

I have no idea if all the RegExes that I have defined actually work the way they are supposed to ... there are 100 different RegEx specifications and Lulu doesn't tell me which one it expects..... `\.*` can mean very different things.

**Comment by xanoni** [comment]
Just looked at the code and it seems LuLu expects proper regexes ....

https://github.com/objective-see/LuLu/blob/dfe48c2a0f7a86361bb1052b3c409330a1ec1b05/LuLu/Extension/Rules.m#L671

Here is Apple's documentation of the syntax: https://developer.apple.com/documentation/foundation/nsregularexpression

**Comment by tomtastic** [comment]
One year and this is still a problem?!
Time for me to unsubscribe from this issue :)

**Comment by ohuf** [comment]
> Just looked at the code and it seems LuLu expects proper regexes .... @ghost have you tried something like `.*api\.netscape\.com.*`?
> 
> https://github.com/objective-see/LuLu/blob/dfe48c2a0f7a86361bb1052b3c409330a1ec1b05/LuLu/Extension/Rules.m#L671
> 
> Here is Apple's documentation of the syntax: https://developer.apple.com/documentation/foundation/nsregularexpression

Hey Linus
Thanks for this comment. Your answer helped me a good deal

so allowing access to all IP address resources in my home LAN would go like this:
![Bildschirmfoto 2021-09-10 um 12 16 38](https://user-images.githubusercontent.com/2392740/132838972-13413423-ba1e-4c42-81ce-9e914c68e731.png)

escaping the IP dots with `\.`
and allowing anything in the last octet with  `.*`

**Comment by xanoni** [comment]
> so allowing access to all IP address resources in my home LAN would go like this:

I think it should work, does it? I would probably prefix `.*` before the `192.` just to be safe, in case it requires that to match things like `http://` (=> `.*192\.168\.1\..*`). If it's not needed, it would just be ignored, because `.*` can be empty. If you test these options, would be great to know how it went.

Regarding contributions ... I also new to this repo, but [based on previously accepted Pull Requests, ](https://github.com/objective-see/LuLu/pulls?q=is%3Apr+is%3Aclosed)it seems you just send a PR and wait until someone accepts it :)

**Comment by xanoni** [comment]
> We have to try this out and we need to create a better user documentation with examples.

100% agree, the documentation for the rule creation functionality is super high level and people have to go through the code to use the app properly. Obviously most people won't do that, so they run settings that don't do what they think they do. Not ideal given this is security critical software.

**Comment by xanoni** [comment]
Summary
===

Okay so if we look at the code we learn that:

- Unless the "regex" box is checked, LuLu only knows the wildcard rule (`*`) and exact string matches ... that means that outside regex rules, the `*` wildcard only works if its standalone and not part of a longer string (e.g., `*.google.com` wouldn't do what we expect) ... this also implies that partial matches are only possible if "regex" is checked (e.g. `google` should match `https://www.google.com/is/cool/maybe.html` — please confirm)
- A lot of our confusion is probably caused by the fact that LuLu differentiates between 1) Hostname, 2) Port, and 3) URL (the latter  can include the hostname), and it checks each of the 3 in isolation ... it appears that the hostname is always an IP, and the domain name ("google.com") would be part of the URL instead — please confirm
- Ports are always string-compared, no regex are possible (except for the string `*`, which matches all ports)

**_My conclusion is that unless we want to match one specific IP address, the answer should always be RegEx .... this way we can check both the IP and the Domain Name. Also, we don't require PERFECT character-by-character matches, some variation is fine (e.g., we can match an IP subnet and we can match domain names instead of just IPs)._**

Disclaimer: didn't _test_ any of the above.

Detailed traffic flow
===

So how does it work? If not allowed or blocked by a higher level rule (e.g., block all traffic), the network traffic eventually hits the point where rules need to be checked. See below.

https://github.com/objective-see/LuLu/blob/dfe48c2a0f7a86361bb1052b3c409330a1ec1b05/LuLu/Extension/FilterDataProvider.m#L347

The actual rule processing starts here:

https://github.com/objective-see/LuLu/blob/dfe48c2a0f7a86361bb1052b3c409330a1ec1b05/LuLu/Extension/Rules.m#L446

LuLu performs the rule matching in 4 steps, but only one of the steps needs to indicate a match for the full rule to apply ("OR"):

1. Are both `ADDRESS=*` and `PORT=*` ? => rule matches
2. If not, is at least `PORT=*` ? => check only `ADDRESS`  (this involves checking of HOST and URL separately, see below)
3. If not, is at least `ADDRESS=*` ? => check only `PORT` _[simple string comparison, no regex supported]_
4. None of the above? => check `ADDRESS` _and_ `PORT` _[regex applies only to ADDRESS]_

The address "checks" are below, for the scenarios 2 and 4 from above. Ports, as mentioned, always need to be perfect matches.

https://github.com/objective-see/LuLu/blob/dfe48c2a0f7a86361bb1052b3c409330a1ec1b05/LuLu/Extension/Rules.m#L646

1. If the rule is a regex, ...
1a. ... check if the hostname at least matches part of the regex (I _believe_ hostname is always an IP address ... not 100% sure) => rule matches
1b. ... check if the full URL (domain including protocol etc.) matches the regex (partial match is fine) => rule matches
2. If the rule is not a regex, check the below ...
2a. ... is the hostname a [perfect match](https://developer.apple.com/documentation/foundation/nsstring/1407803-isequaltostring)?  => rule matches
2b. ... is the full URL a [perfect match](https://developer.apple.com/documentation/foundation/nsstring/1407803-isequaltostring)? => rule matches
3. Otherwise, no match

**Comment by xanoni** [comment]
I'll try to answer the questions in this thread based on my newly gained knowledge.

> Should the value be `^api\.netscape\.com$` or `api\.netscape\.com` (@ghost)

The latter should work, the former probably not, given it is missing protocol / port / resource. LuLu does a simple URI match.

> I tried a regex of `.*\.protonmail\.ch` to match all subdomains, but as per screenshot LuLu fails to match it... (@tomtastic)

LuLu does a primitive string match against the URL. You'd have to specify the IPs manually in a regex.

> I wanted to allow access to my local IP range (192.168.67.0/24), so I added a destination regex of
> 192.168.1..* (@ohuf)

Is this a typo? Shouldn't the regex be `192\.168\.67\.[0-9]+` or similar?

> so allowing access to all IP address resources in my home LAN would go like this:
> ![Bildschirmfoto 2021-09-10 um 12 16 38](https://user-images.githubusercontent.com/2392740/132838972-13413423-ba1e-4c42-81ce-9e914c68e731.png)
> 
> escaping the IP dots with `\.`
> and allowing anything in the last octet with `.*`

Yes this should work. (@ohuf)
